File: leetcode/three_sum_closest.java
public class Solution {
    public int threeSumClosest(int[] num, int target) {
        int result = 0;
        int closest = Integer.MAX_VALUE;
        int len = num.length;
        Arrays.sort(num);

        for (int i = 0; i < len - 2; ++i) {
            int j = i + 1, k = len - 1;
            while (j < k) {
                int sum = num[i] + num[j] + num[k];
                if (Math.abs(sum - target) < closest) {
                    closest = Math.abs(sum - target);
                    result = sum;
                }
                if (sum > target) {
                    --k;
                } else if (sum < target) {
                    ++j;
                } else {
                    return result;
                }
            }
        }
        return result;
    }
}

File: leetcode/three_sum_closest.py
import sys

def three_sum_closest(num, target):
    closest = sys.maxint
    result = 0
    len_num = len(num)
    num.sort()
    for i in xrange(len_num - 2):
        j = i + 1
        k = len_num - 1
        while (j < k):
            sum = num[i] + num[j] + num[k]
            if abs(sum - target) < closest:
                result = sum
                closest = abs(sum - target)
            if sum - target > 0:
                k -= 1
            elif sum - target < 0:
                j += 1
            else:
                return sum
    return sum

assert(three_sum_closest([-1, 2, 1, -4], 1) == 2)
